modules/custom/wim_cvdr/wim_cvdr.module

```diff
@@ -0,0 +1,235 @@
+      sort($items);
+
+      foreach ($items as $item) {
+        $list_items[] = l($item['title'], $item['url'],
+          array('attributes' => array('target' => '_blank')));
+      }
+
+      $block['content'] = theme('item_list', array(
+        'items' => $list_items,
+        'title' => '',
+        'type' => 'ul',
+        'attributes' => array(
+          'class' => array('cvdr-wim', drupal_html_class($subject)),
+        ),
+      ));
+    }
+  }
+
+  // Add JS.
+  drupal_add_js(array(
+    'wim_cvdr' => array(
+      'maxItems' => variable_get('wim_cvdr_max_visible', 5) - 1,
+    ),
+  ), 'setting');
+  drupal_add_js(
+    drupal_get_path('module', 'wim_cvdr') . '/wim_cvdr.js', array(
+      'type' => 'file',
+      'scope' => 'footer',
+    )
+  );
+
+  return $block;
+}
+
+/**
+ * Retrieve the subjects available from the webservice.
+ *
+ * @return array
+ *   containing the subjects.
+ */
+function _wim_cvdr_retrieve_subjects() {
+  $items = array();
+
+  // Use caching to improve performance.
+  if ($cached = cache_get('wim_cvdr_items')) {
+    $items = $cached->data;
+  }
+  else {
+    $xml = _wim_cvdr_read_xml();
+    if ($xml) {
+      $items = _wim_cvdr_process_xml($xml);
+    }
+  }
+
+  return array_keys($items);
+}
+
+/**
+ * Retrieve all items from the webservice.
+ *
+ * @param string $key
+ *   Indicating which block to render.
+ *
+ * @return array
+ *   Containing all items.
+ */
+function _wim_cvdr_retrieve_items($key = '') {
+  $items = array();
+
+  // Use caching to improve performance.
+  if ($cached = cache_get('wim_cvdr_items')) {
+    $items = $cached->data;
+  }
+  else {
+    $xml = _wim_cvdr_read_xml();
+    if ($xml) {
+      $items = _wim_cvdr_process_xml($xml);
+    }
+  }
+
+  // Return specific items if requested.
+  if (!empty($key) && array_key_exists($key, $items)) {
+    return $items[$key];
+  }
+  elseif (!empty($key) && !array_key_exists($key, $items)) {
+    return t('Block could not be rendered.');
+  }
+
+  // Return all items.
+  return $items;
+}
+
+/**
+ * Request the XML from the webservice.
+ */
+function _wim_cvdr_read_xml() {
+  $xml_url = variable_get('wim_cvdr_url');
+
+  // Request the XML from the webservice.
+  $xml_response = drupal_http_request($xml_url, array('timeout' => 10));
+
+  if ($xml_response->code !== '200' || strpos($xml_response->headers['content-type'], 'text/xml') === FALSE) {
+    // Log error and quit.
+    watchdog('WIM CVDR',
+      'Could not retrieve XML. Debug: !array',
+      array('!array' => print_r($xml_response, TRUE)),
+      WATCHDOG_ERROR);
+    return FALSE;
+  }
+
+  // Return the XML data.
+  return $xml_response->data;
+}
+
+/**
+ * Process the XML by creating an array keyed on subject.
+ *
+ * @param string $xml
+ *   The xml in string format.
+ *
+ * @return array
+ *   With items keyed on subject and sorted.
+ */
+function _wim_cvdr_process_xml($xml) {
+  $xml_dom = new DOMDocument();
+  $xml_dom->loadXML($xml);
```

Comment:
The application crashes if our fetched XML is malformed and couldn't be properly loaded/parsed. We should handle this gracefully by logging a helpful error and returning (an empty array).

Reply:
i think we can use following step to check it:
```PHP
 if ($xml === FALSE) {
    // Log error and quit.
    watchdog('WIM CVDR',
      'XML is malformed and couldn\'t be properly loaded/parsed. Errors: !message',
      ['!message' => print_r(libxml_get_errors(), TRUE)],
      WATCHDOG_ERROR);
    libxml_clear_errors();
    return $items;
  }
```

Reply:
As per http://php.net/manual/en/domdocument.loadxml.php the loadXML method returns `TRUE` on success or `FALSE` on failure.

So we probably want something slightly different :)
```php
  $status =  $xml_dom->loadXML($xml);
  // If parsing the XML failed we log an error and stop processing.
  if (!$status) {
    watchdog('WIM CVDR',
      'XML is malformed and couldn\'t be properly loaded/parsed. Errors: !message',
      ['!message' => print_r(libxml_get_errors(), TRUE)],
      WATCHDOG_ERROR);
    libxml_clear_errors();
    return $items;
  }
```